src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Range = Microsoft.VisualStudio.LanguageServer.Protocol.Range;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    internal static class DocumentOutlineHelper
+    {
+        internal static DocumentSymbol[] GetNestedDocumentSymbols(DocumentSymbol[]? documentSymbols)
+        {
+            if (documentSymbols is null || documentSymbols.Length == 0)
+                return Array.Empty<DocumentSymbol>();
+
+            var allSymbols = documentSymbols
+                .SelectMany(x => x.Children)
+                .Concat(documentSymbols)
+                .OrderBy(x => x.Range.Start.Line)
+                .ThenBy(x => x.Range.Start.Character)
+                .ToList();
+
+            return GroupDocumentSymbolTrees(allSymbols)
+                .Select(group => CreateDocumentSymbolTree(group))
+                .ToArray();
+        }
+
+        // Groups a flat list of document symbols into lists containing the symbols of a tree
+        // The first symbol in a list is always the parent (determines the group's position range)
+        private static List<List<DocumentSymbol>> GroupDocumentSymbolTrees(List<DocumentSymbol> allSymbols)
```

Comment:
lots of usages of mutable lists (which makes things difficult to reason about in terms of mutation and therading).  can we move to ImmutableArrays and immutable data types as much as possible in the entire PR>